### PR TITLE
Bug 578871 - prevent UI freeze during UNDO operation + autobuild

### PR DESCRIPTION
**Pull request by jukzi**
While JDT's refactoring correctly tries to interrupt autobuild the UNDO
of the same operation did not. In general all UNDO Operations did not
try to interrupt autobuild.

This adds dependency from ui to resources to make the History operations
become aware of resource scheduling rule.

from https://git.eclipse.org/r/c/platform/eclipse.platform.ui/+/192458

### CODE REVIEW

**Comment by mickaelistria** [comment]
As mentioned on Gerrit, I'm not really convinced that this a proper fix to the right problem. The actual problem is trying to begin a rule in the UI Thread should show some progress dialog. That would cover all cases of that kind without need for the operation to care about the progress report, and instead to focus on actually doing its job.
So I would rather not merge it. I will have more time for development in May and think I can implement a more general solution then.

**Comment by ghost** [comment]
How about submit this and **just revert** if it makes any trouble or when you found a better solution?
(I have doubts you will find a general solution in time)

**Comment by mickaelistria** [comment]
OK, let's do that.

**Comment by mickaelistria** [comment]
I opened #8 about a general solution.

File: bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/operations/OperationHistoryActionHandler.java [hunk 1]
```diff
@@ -298,7 +303,27 @@ public final void run() {
...
+			} else {
+				// prevent UI freeze during AutoBuild as in 
+				// org.eclipse.jdt.internal.ui.refactoring.RefactoringExecutionHelper#perform
+				final IJobManager manager = Job.getJobManager();
+				// pessimistic scheduling rule:
+				final ISchedulingRule rule = ResourcesPlugin.getWorkspace().getRoot(); 
+				try {
+					try {
+						// interrupt autobuild or show Wait/Cancel Dialog:
+						Runnable r = () -> manager.beginRule(rule, null);
```
Comment (iloveeclipse) [inline bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/operations/OperationHistoryActionHandler.java:317]: This change causes a regression. If the Undo task joins on a workspace rule, it hangs now.
So starting a rule here is not OK.
I would propose to revert the change.

Comment (jukzi) [reply bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/operations/OperationHistoryActionHandler.java:317]: > If the Undo task joins on a workspace rule

how to reproduce that?

Comment (iloveeclipse) [reply bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/operations/OperationHistoryActionHandler.java:317]: Have custom `IUndoableOperation` (we have one that extends `AbstractOperation`) that runs operation in a different thread and there takes the workspace rule.

```
    @Override
    public IStatus undo(IProgressMonitor monitor, IAdaptable info) throws ExecutionException {

        IProgressService progressService = PlatformUI.getWorkbench().getProgressService();
        try {
            progressService.run(true, false, new IRunnableWithProgress() {
                @Override
                public void run(IProgressMonitor monitor) throws InvocationTargetException, InterruptedException {
                    ISchedulingRule rule = ResourcesPlugin.getWorkspace().getRoot();
                   Job.getJobManager().beginRule(rule, monitor); // here we hang now
                   // ... do something that needs workspace
                }
            });
        } catch (InvocationTargetException | InterruptedException e) {
        }

        return Status.OK_STATUS;
    }
```

Comment (iloveeclipse) [reply bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/operations/OperationHistoryActionHandler.java:317]: Basically, locking **anything accessible to 3rd party code** before starting undo is not OK, because "undo" in 3rd party code can require the same lock.

Also interesting (I haven't noticed that first), that the org.eclipse.ui.workbench bundle in question had no references to org.eclipse.core.resources bundle before, I think for a good reason.
